File: src/skillenv/plugins.py
from skillenv.envs import Env
from skillenv.lock import add_plugin_record, read_lock


def install_plugin(env: Env, selector: str) -> str:
    config_path = env.root / "config.toml"
    existing = config_path.read_text(encoding="utf-8") if config_path.exists() else ""
    block_header = f'[plugins."{selector}"]'
    if block_header not in existing:
        separator = "\n" if existing.endswith("\n") else "\n\n"
        config_path.write_text(existing + separator + f'{block_header}\nenabled = true\n', encoding="utf-8")
    add_plugin_record(env, name=selector, source=selector)
    return selector


def list_plugins(env: Env) -> list[str]:
    return [plugin["name"] for plugin in read_lock(env)["plugins"]]

File: src/plugins.ts
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { addPluginRecord, readLock } from "./lock.js";

/**
 * Record a plugin selector in the environment's `config.toml` as a
 * `[plugins."<selector>"]` block with `enabled = true`, mirroring the Codex
 * plugin config format.
 */
export function installPlugin(envRoot: string, selector: string): string {
  if (selector.trim().length === 0) {
    throw new Error("plugin selector cannot be empty");
  }
  const configFile = path.join(envRoot, "config.toml");
  const existing = existsSync(configFile) ? readFileSync(configFile, "utf8") : "";
  const blockHeader = `[plugins."${selector}"]`;
  if (!existing.includes(blockHeader)) {
    const separator = existing.endsWith("\n") ? "" : "\n\n";
    writeFileSync(configFile, `${existing}${separator}${blockHeader}\nenabled = true\n`, "utf8");
  }
  addPluginRecord(envRoot, selector, selector);
  return selector;
}

export function listPlugins(envRoot: string): string[] {
  return readLock(envRoot).plugins.map((plugin) => plugin.name);
}
